compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -291,6 +294,16 @@ trait SpaceLogic {
 class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
   import tpd._
 
+  override def intersectAtomTypes(tp1: Type, tp2: Type) = {
+    def isOpen(tp: Type) =
+      !tp.classSymbol.is(Sealed | Final) && !tp.termSymbol.is(Module)
+
+    if ((tp1.classSymbol.is(Trait) || tp2.classSymbol.is(Trait))
+        && isOpen(tp1) && isOpen(tp2))
+      Typ(AndType(tp1, tp2), true)
```

Comment:
Could you elaborate how and-types could be abused here? Judging from the `classSymbol` [implementation](https://github.com/lampepfl/dotty/blob/f009dc0d6d7d93a378da3411bd41bbcb29428db5/compiler/src/dotty/tools/dotc/core/Types.scala#L294), if the type is an and-type, a symbol will be returned only if the and-type is "effectively" a class/trait, which seems to be correct in this case. 

We could potentially simplify the and-type here if that's the case, though.